apiserver/tools.go

```diff
@@ -277,98 +278,31 @@ func (h *toolsDownloadHandler) getToolsForRequest(r *http.Request) (_ io.ReadClo
 	if err != nil {
 		return nil, 0, errors.Trace(err)
 	}
-	binaries, err := ds.AgentBinary().ListAgentBinaries(r.Context())
-	if err != nil {
-		return nil, 0, errors.Trace(err)
-	}
-	sha256 := ""
-	for _, v := range binaries {
-		// TODO(agentbinary): perform this search without listing all cached
-		// agent binaries.
-		if v.Version == vers.Number.String() && v.Arch == vers.Arch {
-			sha256 = v.SHA256
-			break
-		}
-	}
-	if sha256 == "" {
-		// TODO(agentbinary): if the binary is not found, ask simplestreams for
-		// it, then cache it in the model agent binary store. It is possible to
-		// cache in the controller agent binary store if the source stream is
-		// the default.
-		return nil, 0, errors.NotFound
-	}
 
-	// Check the current model's agent binary store first.
-	f, size, err := ds.AgentBinaryStore().GetAgentBinaryForSHA256(r.Context(), sha256)
-	if errors.Is(err, agentbinaryerrors.NotFound) {
-		// Fallback to the controller agent binary store.
-		f, size, err = ds.ControllerAgentBinaryStore().GetAgentBinaryForSHA256(r.Context(), sha256)
+	ctx := r.Context()
+	agentBinaryService := ds.AgentBinary()
+	var agentBinaryStream io.ReadCloser
+	var size int64
+
+	ver := coreagentbinary.Version{
+		Arch:   vers.Arch,
+		Number: vers.Number,
 	}
+
+	hasAgentBinary, err := agentBinaryService.HasAgentBinary(ctx, ver)
 	if err != nil {
-		return nil, 0, errors.Trace(err)
+		return nil, 0, errors.Errorf("checking if agent binary exists: %v", err)
 	}
 
-	return f, size, nil
-}
+	str := coreagentbinary.AgentStream(stream)
+	if hasAgentBinary {
+		agentBinaryStream, size, err = agentBinaryService.GetAgentBinary(ctx, ver) // model
+	} else {
+		agentBinaryStream, size, err = agentBinaryService.GetExternalAgentBinary(ctx, ver, str) // controller, simplestreams
```

Comment:
We need some proper error handling here. Happy to land this PR and follow it up.

Reply:
np, refactored in this pr w thanks!